programs/survival/horse_buddy.sc

```diff
@@ -0,0 +1,306 @@
+  },
+  'weakness' -> {
+    'generic.movement_speed' -> 0.4,
+    'horse.jump_strength' -> 0.4,
+    'generic.max_health' -> 0.4
+  },
+  'regeneration' -> {
+    'generic.movement_speed' -> 0.45,
+    'horse.jump_strength' -> 0.45,
+    'generic.max_health' -> 0.6
+  },
+  'poison' -> {
+    'generic.movement_speed' -> 0.6,
+    'horse.jump_strength' -> 0.6,
+    'generic.max_health' -> 0.2
+  }
+};
+// Accumulate the potion group stat modifiers.
+_effect_stat_mods(effects) -> (
+  reduce(effects,
+    for(pairs(_potion_groups():_),
+      _a:(_:0) = sum(_a:(_:0), _:1,  -0.5);
+    );
+    _a
+  ,
+  { 'generic.movement_speed' -> 0.5,
+    'horse.jump_strength' -> 0.5,
+    'generic.max_health' -> 0.5
+  })
+);
+
+
+// min and max clamps for horse stats.
+_horse_stat_range() -> ({
+  'generic.movement_speed' -> [0.112499997,         0.337499997],
+  'horse.jump_strength' ->    [0.4000000059604645,  1.000000006],
+  'generic.max_health' ->     [15, 46]
+  // 15,32 is vanilla range. 
+  // I just want to allow horses to become more durable against surprise creepers.
+  // Higher than vanilla jump results in horses taking fall damage.
+  // Not much higher than vanilla speed is honestly rediculus, almost unplayable on normal terrain.
+});
+
+
+// keep track of how much training the players horse has done.
+// training mosting meaning using the horse normally.
+global_training = {
+  'uuid' -> null,
+  'food' -> 'wheat',
+  'generic.movement_speed' -> 0,
+  'horse.jump_strength' -> 0,
+  'generic.max_health' -> 0
+};
+
+// Rest the training when ridding a new horse
+// or after rolling for new stats.
+_rest_training(uuid) -> (
+  global_training = {
+    'uuid' -> uuid,
+    'food' -> _random_food(),
+    'generic.movement_speed' -> 0,
+    'horse.jump_strength' -> 0,
+    'generic.max_health' -> 0
+  };
+);
+
+
+// Roll an amount to add or remove from a given stat
+_roll_stat_for(stat, modifier) -> (
+  call(
+    // Adjust movement speed by 10% of vanilla range
+    { 'generic.movement_speed' -> _(x) -> (
+      0.225 * 0.2 * ((x) - rand(1))
+    ),
+    // Adjust jump height by roughly 1/2 blocks
+    'horse.jump_strength' -> _(x) -> (
+      0.25 * ((x) - rand(1))
+    ),
+    // Adjust health by 1/2 Hearts
+    'generic.max_health' -> _(x) -> (
+      3.0 - floor(rand(6)+((x-0.5)*6))
+    )
+  }:stat, modifier)
+);
+
+_clamp(mn, mx, val) -> max(mn, min(mx, val));
+
+
+
+//--- FOOD ---//
+
+_random_food() -> keys(_food_strengths()):(floor(rand(6)));
+
+_food_strengths() -> ({
+  'wheat'->1,
+  'apple'->2,
+  'golden_carrot'->6,
+  'golden_apple'->64,
+  'hay_block'->12
+});
+
+
+//--- ON PLAYER METHODS ---//
+
+__on_player_rides(p, forward, strafe, jumping, sneaking) -> (
+  global_training:'horse.jump_strength' += number(jumping);
+  global_training:'generic.movement_speed' += ceil(forward) + ceil(strafe);
+);
+
+__on_player_interacts_with_entity(p, horse, hand) -> (
+  kind = horse~'type';
+  if(kind == 'horse' || kind == 'donkey',
```

Comment:
does this exclude mules? Or are they just classed as donkeys?

Reply:
added mules